### PR TITLE
Alma: Display descriptions for coded fields while retaining translata…

### PR DESCRIPTION
**Pull request by EreMaijala**
…bility using the codes.

This is a proposal to expand the use of TranslatableString to data elements that contain a code and a translation. I started to think about this when the Alma driver was showing codes for borrowing location information for loans. The codes make sense for translation, but not so much for display.

### CODE REVIEW

**Comment by demiankatz** [comment]
@EreMaijala, without knowing what Alma codes actually look like, I have to ask: should we be using a text domain on these to ensure that the translations can be meaningfully controlled, or is it distinct enough to be safe as-is?

I'm also flagging @betullam for input on this (as well as #1444) in case he has any Alma-specific input to add before I merge.

**Comment by EreMaijala** [comment]
A text domain would make sense, and it probably needs to be configurable since with MultiBackend there can be multiple Alma instances and I suppose they could have different translations for same codes.

**Comment by EreMaijala** [comment]
Another possibility is a configurable prefix like there's in SierraRest driver. I'm not sure which one would be better in this case. We have about 80 ILS's, and the thought of having a separate text domain for each one of them makes me shudder a bit.

**Comment by demiankatz** [comment]
@EreMaijala, a configurable option makes sense to me. If you use a configurable prefix and support the `::` syntax for text domains in the prefix, then you could potentially allow both approaches to work, depending on user preferences.

**Review by betullam — APPROVED** [review]
No objections from my side. I can't comment on text domains as we never had the need to use them.

**Comment by EreMaijala** [comment]
There's now a setting for translation prefix. It can be something simple like "alma_" or a text domain like "Alma::" or even a combination of both. I had to tweak the checkedout template a bit since it already has the hard-coded location_ prefix.

**Comment by demiankatz** [comment]
Thanks, @EreMaijala, though I wonder if there's any risk that the tweaked checkedout template will now cause strings to get translated into something inappropriate, if location names happen to overlap with existing translation strings. (Not sure how likely that is -- but seems not completely impossible). This also makes me wonder where else we have hard-coded `location_` strings -- I haven't checked, but it seems likely this isn't the only example. There's also the https://vufind.org/jira/browse/VUFIND-1236 ticket to think about; I wonder if we ought to make the location prefix globally configurable... and whether such a config option would potentially provide a solution to my other theoretical problem.

I'm a bit short on time today, so I'll try to give this a bit of thought and perhaps make more complete suggestions next week when I can devote more time to the problem!

**Comment by EreMaijala** [comment]
@demiankatz You're right, it's possible to clash without a prefix. I'll think about a better approach. The location_ prefix is only really a problem if you try to specify a text domain in the Alma config, so I could add a check for '::' before prefixing with 'location_', but that feels hacky. Another option would be to let TranslatableString carry the domain separately, but then you'd have to use the location_ prefix in the domain-specific file too.

**Comment by EreMaijala** [comment]
We could also assume that if it's a TranslatableString, it should not be prefixed. Maybe that'd be the cleanest approach.

**Comment by demiankatz** [comment]
@EreMaijala, I think you're right that skipping the prefix if it's a translatable string seems like a pretty safe approach.

Regarding the other issue of eliminating the location_ prefix entirely in the future, I wonder if it might make sense to have a view helper for setting up location strings, which could be made configurable (so by default we keep location_ as the default for backward compatibility, but we could also switch to a text domain, or turn it off entirely). Obviously that's a separate issue -- but just brainstorming while it's on my mind.

**Comment by EreMaijala** [comment]
@demiankatz The view helper felt like such a good idea that I went ahead and added it. That simplified a lot of things apart from GetItemStatuses.

**Comment by demiankatz** [comment]
@EreMaijala, two things:

1.) I wonder if `translateLocation` is unnecessarily specific. Could we accomplish the same thing using something like `translateWithPrefix` so we could potentially reuse this same logic in other situations? Or do you think this is a bad legacy practice, and it's better to compartmentalize it like this?

2.) There's a merge conflict -- can you resolve when time permits?

**Comment by EreMaijala** [comment]
@demiankatz I think that text domains should be encouraged instead of prefixes. The prefixing practice predates support for domains, so needs to be supported for the time being, but also that could be revisited at some point. That's why I'd rather not add a generic method. If you feel that would make sense, would it be better to just add another parameter to the existing translate/transEsc helpers?

**Comment by demiankatz** [comment]
@EreMaijala, I agree that we should not encourage non-text domain prefixes. The question, though, is whether we can come up with a mechanism that addresses prefixes now and is useful for text domains in the future. The other related thread is [VUFIND-1236](https://vufind.org/jira/browse/VUFIND-1236), which talks about eliminating the location prefix; I'd like for our solution today to make that easier to resolve in future with a minimum of fuss.

I think my main concern with this as-is is that it feels excessive to have a new method in the TranslatorAwareTrait (which will thus be included all over the place) as well as a whole new view helper, both of which are only designed to display a single field. On the plus side, this does very nicely encapsulate the actual `location_` prefix in one place, where it can be easily swapped for a text domain in the future... but if we do that, do we then eliminate the need for all of this infrastructure entirely and end up having to rip it back out again anyway?

Adding a fourth parameter to translate/transEsc doesn't feel like the right solution because I think the interaction between the $default and $prefix parameters might be unclear and confusing... not to mention that all the invocations throughout the code would be ugly because of the need to pass in an empty array and null for the second and third parameters.

A separate translateWithPrefix / transEscWithPrefix approach feels more generic, but honestly doesn't appeal much more -- as you say, this might encourage use of what is no longer a good practice, and it potentially requires us to add two new view helpers rather than just one. (We could omit translateWithPrefix if we don't actually use it anywhere, but that feels unbalanced).

One thought, which may not be a solution in itself, but which comes at this from a slightly different angle: what if we had a configuration for the location prefix, so that for now, it remains 'location_' but in future we can change it to 'Location::' to resolve [VUFIND-1236](https://vufind.org/jira/browse/VUFIND-1236)? (And then you can locally set it back to 'location_' temporarily or permanently if that's easier for your needs than switching to a text domain). I think this approach would make it impossible to define a location-specific method in TranslatorAwareTrait since we would need some kind of non-static container to fetch the configuration properly... but since I don't like having that method in that trait, that's not necessarily a bad thing. We could then add a more generic prefix-oriented support method to the trait to make implementation of this functionality easier, and then I'm wondering if things just fall into place from that seed. Or maybe I'm on the wrong track -- anyway, food for thought!

Sorry for the verbose reply -- something just doesn't quite "feel right" here yet, and I'm hoping we can make it fit together just a little bit better if we take our time.

**Comment by EreMaijala** [comment]
@demiankatz I renamed the functions. While this doesn't directly allow transitioning to a text domain, at least the places where the location prefix is used are easy to find. If we want to make the prefix configurable, I'd rather do that separately. Would this work for you?

File: module/VuFind/src/VuFind/AjaxHandler/GetItemStatuses.php [hunk 1]
```diff
@@ -139,9 +139,11 @@ protected function translateList($transPrefix, $list)
     {
         $transList = [];
         foreach ($list as $current) {
-            $transList[] = $this->translate(
-                $transPrefix . $current, [], $current
-            );
+            $default = $current;
+            if (!($current instanceof \VuFind\I18n\TranslatableString)) {
```
Comment (demiankatz) [inline module/VuFind/src/VuFind/AjaxHandler/GetItemStatuses.php]: Can we simplify the code changes in this file by using translateWithPrefix now that we have defined that method? Should we move the TranslatableString awareness into that function as well?

**Review by demiankatz — COMMENTED** [review]
@EreMaijala, I agree that the configurable prefix is a separate concern that we can deal with elsewhere. Just one last question before I run tests one last time and (hopefully) merge this....

File: module/VuFind/src/VuFind/AjaxHandler/GetItemStatuses.php (hunk 1 above)
Comment (EreMaijala) [reply module/VuFind/src/VuFind/AjaxHandler/GetItemStatuses.php]: Indeed we can. :) Done.

**Comment by demiankatz** [comment]
Tests all pass -- just merged!

**Pull request merged**